tools/walletextension/persistence/persistence.go

```diff
@@ -0,0 +1,164 @@
+package persistence
+
+import (
+	"encoding/csv"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/crypto/ecies"
+	"github.com/obscuronet/go-obscuro/go/common/log"
+	"github.com/obscuronet/go-obscuro/go/rpc"
+)
+
+const (
+	obscuroDirName           = ".obscuro"
+	persistenceFileName      = "wallet_extension_persistence"
+	persistenceNumComponents = 4
+	persistenceIdxHost       = 0
+	persistenceIdxAccount    = 1
+	persistenceIdxViewingKey = 2
+	persistenceIdxSignedKey  = 3
+)
+
+// Persistence handles the persistence of viewing keys.
+type Persistence struct {
+	filePath string // The path of the file used to store the submitted viewing keys
+	hostAddr string // The address of the host the keys are being persisted for
+}
+
+func NewPersistence(hostAddr string, persistenceFilePath string) *Persistence {
+	// Sets up the persistence file and returns its path. Defaults to the user's home directory if the path is empty.
+	if persistenceFilePath == "" {
+		homeDir, err := os.UserHomeDir()
+		if err != nil {
+			panic("cannot create persistence file as user's home directory is not defined")
+		}
+		obscuroDir := filepath.Join(homeDir, obscuroDirName)
+		err = os.MkdirAll(obscuroDir, 0o777)
+		if err != nil {
+			panic(fmt.Sprintf("could not create %s directory in user's home directory", obscuroDirName))
+		}
+
+		persistenceFilePath = filepath.Join(obscuroDir, persistenceFileName)
+	}
+
+	_, err := os.OpenFile(persistenceFilePath, os.O_CREATE|os.O_RDONLY, 0o644)
+	if err != nil {
+		panic(fmt.Sprintf("could not create persistence file. Cause: %s", err))
+	}
+
+	return &Persistence{
+		filePath: persistenceFilePath,
+		hostAddr: hostAddr,
+	}
+}
+
+// PersistViewingKey persists a viewing key to disk.
+func (p *Persistence) PersistViewingKey(viewingKey *rpc.ViewingKey) {
+	viewingPrivateKeyBytes := crypto.FromECDSA(viewingKey.PrivateKey.ExportECDSA())
+
+	record := []string{
+		p.hostAddr,
+		viewingKey.Account.Hex(),
+		// We encode the bytes as hex to ensure there are no unintentional line breaks to make parsing the file harder.
+		hex.EncodeToString(viewingPrivateKeyBytes),
+		hex.EncodeToString(viewingKey.SignedKey),
+	}
+
+	persistenceFile, err := os.OpenFile(p.filePath, os.O_APPEND|os.O_WRONLY, 0o644)
```

Comment:
Does this lock the file so that another WE process would need to wait until the lock is released before it wrote to the file etc?

Reply:
AFAIK, based on my understanding of `O_APPEND`, each line written should be atomic. Should there's no risk of lines being overwritten or garbled.